Examples/CaseStudies/SwiftUICaseStudies/02-Effects-WebSocket.swift

```diff
@@ -0,0 +1,374 @@
+        return false
+      }
+    }
+  }
+
+  var cancel: (AnyHashable, URLSessionWebSocketTask.CloseCode, Data?) -> Effect<Never, Never>
+  var open: (AnyHashable, URL, [String]) -> Effect<Action, Never>
+  var receive: (AnyHashable) -> Effect<Message, NSError>
+  var send: (AnyHashable, URLSessionWebSocketTask.Message) -> Effect<NSError?, Never>
+  var sendPing: (AnyHashable) -> Effect<NSError?, Never>
+}
+
+extension WebSocketClient {
+  static let live = WebSocketClient(
+    cancel: { id, closeCode, reason in
+      .fireAndForget {
+        dependencies[id]?.task.cancel(with: closeCode, reason: reason)
+        dependencies[id]?.subscriber.send(completion: .finished)
+        dependencies[id] = nil
+      }
+    },
+    open: { id, url, protocols in
+      Effect.run { subscriber in
+        let delegate = WebSocketDelegate(
+          didBecomeInvalidWithError: {
+            subscriber.send(.didBecomeInvalidWithError($0 as NSError?))
+          },
+          didClose: {
+            subscriber.send(.didClose(code: $0, reason: $1))
+          },
+          didCompleteWithError: {
+            subscriber.send(.didCompleteWithError($0 as NSError?))
+          },
+          didOpenWithProtocol: {
+            subscriber.send(.didOpenWithProtocol($0))
+          })
+        let session = URLSession(configuration: .default, delegate: delegate, delegateQueue: nil)
+        let task = session.webSocketTask(with: url, protocols: protocols)
+        task.resume()
+        dependencies[id] = Dependencies(delegate: delegate, subscriber: subscriber, task: task)
+        return AnyCancellable {
+          task.cancel(with: .normalClosure, reason: nil)
+          dependencies[id]?.subscriber.send(completion: .finished)
+          dependencies[id] = nil
+        }
+      }
+    },
+    receive: { id in
+      .future { callback in
+        dependencies[id]?.task.receive { result in
+          switch result.map(Message.init) {
+          case let .success(.some(message)):
+            callback(.success(message))
+          case .success(.none):
+            callback(.failure(NSError.init(domain: "co.pointfree", code: 1)))
+          case let .failure(error):
+            callback(.failure(error as NSError))
+          }
+        }
+      }
+    },
+    send: { id, message in
+      .future { callback in
+        dependencies[id]?.task.send(message) { error in
+          callback(.success(error as NSError?))
+        }
+      }
+    },
+    sendPing: { id in
+      .future { callback in
+        dependencies[id]?.task.sendPing { error in
+          callback(.success(error as NSError?))
+        }
+      }
+    })
+}
+
+private var dependencies: [AnyHashable: Dependencies] = [:]
+private struct Dependencies {
+  let delegate: URLSessionWebSocketDelegate
+  let subscriber: Effect<WebSocketClient.Action, Never>.Subscriber
+  let task: URLSessionWebSocketTask
+}
+
+class WebSocketDelegate: NSObject, URLSessionWebSocketDelegate {
```

Comment:
Can this be private?
```suggestion
private class WebSocketDelegate: NSObject, URLSessionWebSocketDelegate {
```